mm0-rs/src/elab/lisp/debug.rs

```diff
@@ -0,0 +1,1465 @@
+        ron.add_field("args", args.as_ref().dbg_aux(fe, p));
+        ron.to_doc(p)
+      }
+
+      crate::elab::environment::ProofNode::Unfold { term, args, res } => {
+        let mut ron = RonStruct::new();
+        ron.add_name("ProofNode::Unfold");
+        ron.add_field("term", term.dbg_aux(fe, p));
+        ron.add_field("args", args.as_ref().dbg_aux(fe, p));
+        ron.add_field("res", res.as_ref().dbg_aux(fe, p));
+        ron.to_doc(p)
+      }
+    }
+  }
+}
+
+impl EnvDebug for crate::elab::environment::ExprNode {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonTuple::new();
+    ron.add_name("ExprNode");
+    match self {
+      crate::elab::environment::ExprNode::Ref(n) => {
+        ron.add_name("Ref");
+        ron.add_field(n.to_string());
+      }
+      crate::elab::environment::ExprNode::Dummy(atom_id, sort_id) => {
+        ron.add_name("Dummy");
+        ron.add_field(atom_id.dbg_aux(fe, p));
+        ron.add_field(sort_id.dbg_aux(fe, p));
+      }
+      crate::elab::environment::ExprNode::App(term_id, nodes) => {
+        ron.add_name("App");
+        ron.add_field(term_id.dbg_aux(fe, p));
+        ron.add_field(nodes.dbg_aux(fe, p));
+      }
+    }
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for AtomData {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+      let mut ron = RonStruct::new();
+      ron.add_name("AtomData");
+      ron.add_field("name", self.name.to_string());
+      ron.add_field("sort", self.sort.dbg_aux(fe, p));
+      ron.add_field("lisp", self.lisp.dbg_aux(fe, p));
+      ron.add_field("graveyard", self.graveyard.as_deref().dbg_aux(fe, p));
+      ron.to_doc(p)
+  }
+
+
+}
+
+impl EnvDebug for AtomID {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    match self {
+      AtomID(n) => {
+        let atom_data = &fe.data[*self];
+        let mut ron = RonStruct::new();
+        ron.add_name("AtomData");
+        ron.add_field("atom_id", n.to_string());
+        ron.add_field("name", atom_data.name.to_string());
+        ron.add_field("sort", atom_data.sort.dbg_aux(fe, p));
+        ron.add_field("lisp", atom_data.lisp.dbg_aux(fe, p));
+        ron.add_field("graveyard", atom_data.graveyard.as_deref().dbg_aux(fe, p));
+        ron.to_doc(p)
+      }
+    }
+  }
+}
+
+impl EnvDebug for crate::mmc::parser::Invariant {
+  fn dbg_aux<'a>(
+    &self, 
+    fe : FormatEnv<'a>,
+    p : &mut Printer<'a>, 
+  ) -> DocPtr<'a> {
+    let mut ron = RonStruct::new();
+    ron.add_name("Invariant");
+    ron.add_field("name", self.name.dbg_aux(fe, p));
+    ron.add_field("ghost", self.ghost.dbg_aux(fe, p));
+    ron.add_field("ty", self.ty.dbg_aux(fe, p));
+    ron.add_field("val", self.val.dbg_aux(fe, p));
+    ron.to_doc(p)
+  }
+}
+
+impl EnvDebug for crate::mmc::parser::Pattern {
```

Comment:
Some of these debug implementations seem better suited to the individual files that introduce them. The choice of what to put where is a bit of a judgment call, but see the implementations of the `Remap` trait for inspiration. The basic types are implemented with the introduction of the trait, as well as some types that are already required as dependencies for the trait itself, but other types have their implementations elsewhere. In this particular case, I'm trying to minimize references to the `mmc` module from outside it, because it will possibly turn into a separate crate with dynamic linkage, so I would suggest putting these implementations after the declaration of the enum.